# We need the following libraries to work

In [1]:
# !pip install scikit-learn pandas matplotlib xlrd

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [3]:
df = pd.read_excel("./default of credit card clients.xls", index_col=[0])

In [4]:
df.head()

X1   X2         X3        X4   X5     X6     X7     X8     X9   
ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
1       20000    2          2         1   24      2      2     -1     -1   
2      120000    2          2         2   26     -1      2      0      0   
3       90000    2          2         2   34      0      0      0      0   
4       50000    2          2         1   37      0      0      0      0   

      X10  ...        X15        X16        X17       X18       X19       X20   
ID  PAY_5  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
1      -2  ...          0          0          0         0       689         0   
2       0  ...       3272       3455       3261         0      1000      1000   
3       0  ...      14331      14948      15549      1518      1500      1000   
4       0  ...      28314      28959      29547      2000      2019      1200   

         X21       X22       X23                           Y  
ID  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
1          0         0         0                           1  
2       1000         0      2000                           1  
3       1000      1000      5000                           0  
4       1100      1069      1000                           0  

[5 rows x 24 columns]

# Reading the description of the problem, we know that

    X1 = Credit Amount given
    X6 = Repayment status in September
    X12 = Bill amount in September
    X18 = Amount paid in September
    
    Y = default payement next month
    

##### We will model this with the following equation

$\hat Y $ $= X1 * W1 + X6 * W2 + X12 * W3 + X18 * W4 + b $

This is a linear regression model. We will use sklearn linear_model

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [6]:
data_slice = df[['X1', 'X6', 'X12', 'X18', 'Y']].copy()
for col in data_slice.columns:
    data_slice[col] = pd.to_numeric(data_slice[col], errors='coerce')
data_slice.dropna(inplace=True)

In [7]:
datax = data_slice[['X1', 'X6', 'X12', 'X18']]
datay = data_slice[['Y']]

In [8]:
train_x, test_x, train_y, test_y = train_test_split(datax, datay, test_size=0.2, random_state=4215)

In [9]:
train_x.head()

X1   X6       X12     X18
19729  170000.0 -1.0    6699.0  2999.0
10326   30000.0  0.0   29336.0  1800.0
26935  210000.0  2.0  196426.0  8405.0
2884   280000.0 -2.0     593.0  1777.0
17781  100000.0 -1.0    2178.0  7300.0

In [10]:
train_y.head()

Y
19729  0.0
10326  1.0
26935  1.0
2884   0.0
17781  0.0

In [11]:
linear_model = LinearRegression()

In [12]:
linear_model.fit(train_x, train_y)

LinearRegression()

In [13]:
print(f"{np.around( 100 * linear_model.score(test_x, test_y) , 2)}%")

10.39%


In [14]:
print(linear_model.coef_)

[[-1.40995834e-07  1.20565162e-01 -3.73785792e-07 -7.22345074e-07]]


# Since the accuracy is not great (only 10%) we will create a classification model to do the prediction

In [15]:
from sklearn.neighbors import KNeighborsClassifier

In [16]:
knn_model = KNeighborsClassifier()

In [17]:
knn_model.fit(train_x, train_y)

C:\PythonProjects\xcc_DL\Introduction\credit fraurd detection\venv\lib\site-packages\sklearn\neighbors\_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


KNeighborsClassifier()

In [18]:
print(f"{np.around( 100 * knn_model.score(test_x, test_y) , 2)}%")

74.3%


# The accuracy increased from 10% to 74% by changing the model. If we use more features, the accuracy might increase further. 